src/x509/verify.ts

```diff
@@ -0,0 +1,202 @@
+      prev.length < curr.length ? prev : curr
+    );
+
+    // Construct chain from shortest path
+    return [leafCert, ...path];
+  }
+
+  // Recursively build all possible paths from the leaf to the root
+  private buildPaths(certificate: x509Certificate): x509Certificate[][] {
+    const paths = [];
+    const issuers = this.findIssuer(certificate);
+
+    if (issuers.length === 0) {
+      throw new CertificateChainVerificationError(
+        'No valid certificate path found'
+      );
+    }
+
+    for (let i = 0; i < issuers.length; i++) {
+      const issuer = issuers[i];
+
+      // Base case - issuer is self
+      if (issuer.equals(certificate)) {
+        paths.push([certificate]);
+        continue;
+      }
+
+      // Recursively build path for the issuer
+      const subPaths = this.buildPaths(issuer);
+
+      // Construct paths by appending the issuer to each subpath
+      for (let j = 0; j < subPaths.length; j++) {
+        paths.push([issuer, ...subPaths[j]]);
+      }
+    }
+
+    return paths;
+  }
+
+  // Return all possible issuers for the given certificate
+  private findIssuer(certificate: x509Certificate): x509Certificate[] {
+    let issuers: x509Certificate[] = [];
+    let keyIdentifier: Buffer | undefined;
+
+    // Exit early if the certificate is self-signed
+    if (certificate.subject.equals(certificate.issuer)) {
+      if (certificate.verify()) {
+        return [certificate];
+      }
+    }
+
+    // If the certificate has an authority key identifier, use that
+    // to find the issuer
+    if (certificate.extAuthorityKeyID) {
+      keyIdentifier = certificate.extAuthorityKeyID.keyIdentifier;
+
+      // TODO: Add support for authorityCertIssuer/authorityCertSerialNumber
+      // though Fulcio doesn't appear to use these
+    }
+
+    // Find possible issuers by comparing the authorityKeyID/subjectKeyID
+    // or issuer/subject. Potential issuers are added to the result array.
+    this.localCerts.forEach((possibleIssuer) => {
+      if (keyIdentifier) {
+        if (possibleIssuer.extSubjectKeyID) {
+          if (
+            possibleIssuer.extSubjectKeyID.keyIdentifier.equals(keyIdentifier)
+          ) {
+            issuers.push(possibleIssuer);
+          }
+          return;
+        }
+      }
+
+      // Fallback to comparing certificate issuer and subject if
+      // subjectKey/authorityKey extensions are not present
+      if (possibleIssuer.subject.equals(certificate.issuer)) {
+        issuers.push(possibleIssuer);
+      }
+    });
+
+    // Remove any issuers which fail to verify the certificate
+    issuers = issuers.filter((issuer) => {
+      try {
+        return certificate.verify(issuer);
+      } catch (ex) {
+        return false;
+      }
+    });
+
+    return issuers;
+  }
+
+  private checkPath(path: x509Certificate[]): void {
+    if (path.length < 2) {
+      throw new CertificateChainVerificationError(
+        'Certificate chain must contain at least two certificates'
```

Comment:
Worth adding a test case for this?

Reply:
Given the way the certificate chain is assembled in the code above, it's actually not possible to arrive at this point with fewer than two certificates in the chain (other errors would be thrown before you arrive here). I considered removing this check, but I sorta like that it makes explicit a requirement that is difficult to see otherwise (I was thinking about it from the perspective of someone auditing this code).

Happy to delete if you think we should ditch it.

Reply:
Ah yes this makes sense, seems ok to leave it in 👍